File: app/src/main/java/io/beldex/bchat/preferences/NotificationsPreferenceFragment.java
package io.beldex.bchat.preferences;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.media.Ringtone;
import android.media.RingtoneManager;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import android.provider.Settings;
import androidx.annotation.Nullable;
import androidx.preference.ListPreference;
import androidx.preference.Preference;
import android.text.TextUtils;

import io.beldex.bchat.ApplicationContext;
import io.beldex.bchat.components.SwitchPreferenceCompat;
import io.beldex.bchat.notifications.NotificationChannels;
import com.beldex.libbchat.utilities.TextSecurePreferences;

import io.beldex.bchat.R;

import static android.app.Activity.RESULT_OK;

public class NotificationsPreferenceFragment extends ListSummaryPreferenceFragment {

  @SuppressWarnings("unused")
  private static final String TAG = NotificationsPreferenceFragment.class.getSimpleName();

  @Override
  public void onCreate(Bundle paramBundle) {
    super.onCreate(paramBundle);

    // Set up FCM toggle
    String fcmKey = "pref_key_use_fcm";
    ((SwitchPreferenceCompat)findPreference(fcmKey)).setChecked(TextSecurePreferences.isUsingFCM(getContext()));
    this.findPreference(fcmKey)
      .setOnPreferenceChangeListener((preference, newValue) -> {
        TextSecurePreferences.setIsUsingFCM(getContext(), (boolean) newValue);
        ApplicationContext.getInstance(getContext()).registerForFCMIfNeeded(true);
        return true;
      });

    Preference ledBlinkPref = this.findPreference(TextSecurePreferences.LED_BLINK_PREF);

    if (NotificationChannels.supported()) {
      ledBlinkPref.setVisible(false);
      TextSecurePreferences.setNotificationRingtone(getContext(), NotificationChannels.getMessageRingtone(getContext()).toString());
      TextSecurePreferences.setNotificationVibrateEnabled(getContext(), NotificationChannels.getMessageVibrate(getContext()));

    } else {
      ledBlinkPref.setOnPreferenceChangeListener(new ListSummaryListener());
      initializeListSummary((ListPreference) ledBlinkPref);
    }

    this.findPreference(TextSecurePreferences.LED_COLOR_PREF)
        .setOnPreferenceChangeListener(new LedColorChangeListener());
    this.findPreference(TextSecurePreferences.RINGTONE_PREF)
        .setOnPreferenceChangeListener(new RingtoneSummaryListener());
    //New Line
 /*   this.findPreference(TextSecurePreferences.NOTIFICATION_PREF)
            .setOnPreferenceChangeListener(new AllNotificationsListener());*/
    this.findPreference(TextSecurePreferences.REPEAT_ALERTS_PREF)
        .setOnPreferenceChangeListener(new ListSummaryListener());
    this.findPreference(TextSecurePreferences.NOTIFICATION_PRIVACY_PREF)
        .setOnPreferenceChangeListener(new NotificationPrivacyListener());
    this.findPreference(TextSecurePreferences.NOTIFICATION_PRIORITY_PREF)
        .setOnPreferenceChangeListener(new ListSummaryListener());
    this.findPreference(TextSecurePreferences.VIBRATE_PREF)
        .setOnPreferenceChangeListener((preference, newValue) -> {
          NotificationChannels.updateMessageVibrate(getContext(), (boolean) newValue);
          return true;
        });

    this.findPreference(TextSecurePreferences.RINGTONE_PREF)
        .setOnPreferenceClickListener(preference -> {
          Uri current = TextSecurePreferences.getNotificationRingtone(getContext());

          Intent intent = new Intent(RingtoneManager.ACTION_RINGTONE_PICKER);
          intent.putExtra(RingtoneManager.EXTRA_RINGTONE_SHOW_DEFAULT, true);
          intent.putExtra(RingtoneManager.EXTRA_RINGTONE_SHOW_SILENT, true);
          intent.putExtra(RingtoneManager.EXTRA_RINGTONE_TYPE, RingtoneManager.TYPE_NOTIFICATION);
          intent.putExtra(RingtoneManager.EXTRA_RINGTONE_DEFAULT_URI, Settings.System.DEFAULT_NOTIFICATION_URI);
          intent.putExtra(RingtoneManager.EXTRA_RINGTONE_EXISTING_URI, current);

          startActivityForResult(intent, 1);

          return true;
        });

    initializeListSummary((ListPreference) findPreference(TextSecurePreferences.LED_COLOR_PREF));
    initializeListSummary((ListPreference) findPreference(TextSecurePreferences.REPEAT_ALERTS_PREF));
    this.findPreference(TextSecurePreferences.NOTIFICATION_PRIVACY_PREF)
            .setOnPreferenceClickListener(preference -> {
              ListPreference listPreference = (ListPreference) preference;
              listPreference.setDialogMessage(R.string.preferences_notifications__show);
              ListPreferenceDialog list = new ListPreferenceDialog(listPreference, () -> {
                initializeListSummary((ListPreference) findPreference(TextSecurePreferences.NOTIFICATION_PRIVACY_PREF));
                return null;
              });
              list.show(getChildFragmentManager(), "ListPreferenceDialog");
              list.setCancelable(false);
              return true;
            });
    initializeListSummary((ListPreference) findPreference(TextSecurePreferences.NOTIFICATION_PRIVACY_PREF));

    if (NotificationChannels.supported()) {
      this.findPreference(TextSecurePreferences.NOTIFICATION_PRIORITY_PREF)
          .setOnPreferenceClickListener(preference -> {
            Intent intent = new Intent(Settings.ACTION_CHANNEL_NOTIFICATION_SETTINGS);
            intent.putExtra(Settings.EXTRA_CHANNEL_ID, NotificationChannels.getMessagesChannel(getContext()));
            intent.putExtra(Settings.EXTRA_APP_PACKAGE, getContext().getPackageName());
            startActivity(intent);
            return true;
          });
    } else {
      initializeListSummary((ListPreference) findPreference(TextSecurePreferences.NOTIFICATION_PRIORITY_PREF));
    }

    initializeRingtoneSummary(findPreference(TextSecurePreferences.RINGTONE_PREF));
    initializeMessageVibrateSummary((SwitchPreferenceCompat)findPreference(TextSecurePreferences.VIBRATE_PREF));

    //New Line
    /*initializeAllNotifications((SwitchPreferenceCompat)findPreference(TextSecurePreferences.NOTIFICATION_PREF));*/
  }

  @Override
  public void onCreatePreferences(@Nullable Bundle savedInstanceState, String rootKey) {
    addPreferencesFromResource(R.xml.preferences_notifications);
  }

  @Override
  public void onActivityResult(int requestCode, int resultCode, Intent data) {
    if (requestCode == 1 && resultCode == RESULT_OK && data != null) {
      Uri uri = data.getParcelableExtra(RingtoneManager.EXTRA_RINGTONE_PICKED_URI);

      if (Settings.System.DEFAULT_NOTIFICATION_URI.equals(uri)) {
        NotificationChannels.updateMessageRingtone(getContext(), uri);
        TextSecurePreferences.removeNotificationRingtone(getContext());
      } else {
        uri = uri == null ? Uri.EMPTY : uri;
        NotificationChannels.updateMessageRingtone(getContext(), uri);
        TextSecurePreferences.setNotificationRingtone(getContext(), uri.toString());
      }

      initializeRingtoneSummary(findPreference(TextSecurePreferences.RINGTONE_PREF));

      //New Line
      /*initializeAllNotifications((SwitchPreferenceCompat)findPreference(TextSecurePreferences.NOTIFICATION_PREF));*/
    }
  }

  private class RingtoneSummaryListener implements Preference.OnPreferenceChangeListener {
    @Override
    public boolean onPreferenceChange(Preference preference, Object newValue) {
      Uri value = (Uri) newValue;

      if (value == null || TextUtils.isEmpty(value.toString())) {
        preference.setSummary(R.string.preferences__silent);
      } else {
        Ringtone tone = RingtoneManager.getRingtone(getActivity(), value);

        if (tone != null) {
          preference.setSummary(tone.getTitle(getActivity()));
        }
      }

      return true;
    }
  }

  private void initializeRingtoneSummary(Preference pref) {
    RingtoneSummaryListener listener = (RingtoneSummaryListener) pref.getOnPreferenceChangeListener();
    Uri                     uri      = TextSecurePreferences.getNotificationRingtone(getContext());

    listener.onPreferenceChange(pref, uri);
  }

  private void initializeMessageVibrateSummary(SwitchPreferenceCompat pref) {
    pref.setChecked(TextSecurePreferences.isNotificationVibrateEnabled(getContext()));
  }

  //New Line
 /* private void initializeAllNotifications(SwitchPreferenceCompat pref) {
    AllNotificationsListener listener = (AllNotificationsListener) pref.getOnPreferenceChangeListener();
    Boolean                     value      = TextSecurePreferences.isNotificationsEnabled(getContext());

    listener.onPreferenceChange(pref, value);
  }
  private class AllNotificationsListener implements Preference.OnPreferenceChangeListener {
    @Override
    public boolean onPreferenceChange(Preference preference, Object newValue) {
      preference.setEnabled(newValue.toString().equals("true"));
      return true;
    }
  }*/

  public static CharSequence getSummary(Context context) {
    final int onCapsResId   = R.string.ApplicationPreferencesActivity_On;
    final int offCapsResId  = R.string.ApplicationPreferencesActivity_Off;

    return context.getString(TextSecurePreferences.isNotificationsEnabled(context) ? onCapsResId : offCapsResId);
  }

  private class NotificationPrivacyListener extends ListSummaryListener {
    @SuppressLint("StaticFieldLeak")
    @Override
    public boolean onPreferenceChange(Preference preference, Object value) {
      new AsyncTask<Void, Void, Void>() {
        @Override
        protected Void doInBackground(Void... params) {
          ApplicationContext.getInstance(getActivity()).messageNotifier.updateNotification(getActivity());
          return null;
        }
      }.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR);

      return super.onPreferenceChange(preference, value);
    }
  }

  @SuppressLint("StaticFieldLeak")
  private class LedColorChangeListener extends ListSummaryListener {
    @Override
    public boolean onPreferenceChange(Preference preference, Object value) {
      if (NotificationChannels.supported()) {
        new AsyncTask<Void, Void, Void>() {
          @Override
          protected Void doInBackground(Void... voids) {
            NotificationChannels.updateMessagesLedColor(getActivity(), (String) value);
            return null;
          }
        }.execute();
      }
      return super.onPreferenceChange(preference, value);
    }
  }
}

File: app/src/main/java/io/beldex/bchat/preferences/NotificationsPreferenceFragment.kt
package io.beldex.bchat.preferences
import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.media.RingtoneManager
import android.net.Uri
import android.os.AsyncTask
import android.os.Bundle
import android.provider.Settings
import android.text.TextUtils
import androidx.lifecycle.lifecycleScope
import androidx.preference.ListPreference
import androidx.preference.Preference
import com.beldex.libbchat.utilities.TextSecurePreferences
import com.beldex.libbchat.utilities.TextSecurePreferences.Companion.isNotificationsEnabled
import dagger.hilt.android.AndroidEntryPoint
import io.beldex.bchat.ApplicationContext
import io.beldex.bchat.R
import io.beldex.bchat.components.SwitchPreferenceCompat
import io.beldex.bchat.notifications.NotificationChannels
import io.beldex.bchat.notifications.PushRegistry
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import javax.inject.Inject
@AndroidEntryPoint
class NotificationsPreferenceFragment : ListSummaryPreferenceFragment() {
    @Inject
    lateinit var pushRegistry: PushRegistry
    @Inject
    lateinit var prefs: TextSecurePreferences
    override fun onCreate(paramBundle: Bundle?) {
        super.onCreate(paramBundle)
        // Set up FCM toggle
        val fcmKey = "pref_key_use_fcm"
        val fcmPreference: SwitchPreferenceCompat = findPreference(fcmKey)!!
        fcmPreference.isChecked = prefs.isPushEnabled()
        fcmPreference.setOnPreferenceChangeListener { _: Preference, newValue: Any ->
            prefs.setPushEnabled(newValue as Boolean)
            val job = pushRegistry.refresh(true)
            fcmPreference.isEnabled = false
            lifecycleScope.launch(Dispatchers.IO) {
                job.join()
                withContext(Dispatchers.Main) {
                    fcmPreference.isEnabled = true
                }
            }
            true
        }
        val ledBlinkPref = findPreference<Preference>(TextSecurePreferences.LED_BLINK_PREF)
        ledBlinkPref!!.isVisible = false
        prefs.setNotificationRingtone(
            NotificationChannels.getMessageRingtone(requireContext()).toString()
        )
        prefs.setNotificationVibrateEnabled(
            NotificationChannels.getMessageVibrate(requireContext())
        )
        findPreference<Preference>(TextSecurePreferences.LED_COLOR_PREF)!!.onPreferenceChangeListener = LedColorChangeListener()
        findPreference<Preference>(TextSecurePreferences.RINGTONE_PREF)!!.onPreferenceChangeListener = RingtoneSummaryListener()
        //New Line
        /*   this.findPreference(TextSecurePreferences.NOTIFICATION_PREF)
            .setOnPreferenceChangeListener(new AllNotificationsListener());*/
        findPreference<Preference>(TextSecurePreferences.REPEAT_ALERTS_PREF)!!.onPreferenceChangeListener = ListSummaryListener()
        findPreference<Preference>(TextSecurePreferences.NOTIFICATION_PRIVACY_PREF)!!.onPreferenceChangeListener = NotificationPrivacyListener()
        findPreference<Preference>(TextSecurePreferences.NOTIFICATION_PRIORITY_PREF)!!.onPreferenceChangeListener = ListSummaryListener()
        findPreference<Preference>(TextSecurePreferences.VIBRATE_PREF)!!.onPreferenceChangeListener =
            Preference.OnPreferenceChangeListener { _: Preference?, newValue: Any ->
                NotificationChannels.updateMessageVibrate(requireContext(), newValue as Boolean)
                true
            }
        findPreference<Preference>(TextSecurePreferences.RINGTONE_PREF)!!.onPreferenceClickListener =
            Preference.OnPreferenceClickListener {
                val current = prefs.getNotificationRingtone()
                val intent = Intent(RingtoneManager.ACTION_RINGTONE_PICKER)
                intent.putExtra(RingtoneManager.EXTRA_RINGTONE_SHOW_DEFAULT, true)
                intent.putExtra(RingtoneManager.EXTRA_RINGTONE_SHOW_SILENT, true)
                intent.putExtra(
                    RingtoneManager.EXTRA_RINGTONE_TYPE,
                    RingtoneManager.TYPE_NOTIFICATION
                )
                intent.putExtra(
                    RingtoneManager.EXTRA_RINGTONE_DEFAULT_URI,
                    Settings.System.DEFAULT_NOTIFICATION_URI
                )
                intent.putExtra(RingtoneManager.EXTRA_RINGTONE_EXISTING_URI, current)
                startActivityForResult(intent, 1)
                true
            }
        initializeListSummary(findPreference<Preference>(TextSecurePreferences.LED_COLOR_PREF) as ListPreference?)
        initializeListSummary(findPreference<Preference>(TextSecurePreferences.REPEAT_ALERTS_PREF) as ListPreference?)
        findPreference<Preference>(TextSecurePreferences.NOTIFICATION_PRIVACY_PREF)!!.onPreferenceClickListener =
            Preference.OnPreferenceClickListener { preference: Preference ->
                val listPreference = preference as ListPreference
                listPreference.setDialogMessage(R.string.preferences_notifications__show)
                ListPreferenceDialog(listPreference) {
                    initializeListSummary(findPreference(TextSecurePreferences.NOTIFICATION_PRIVACY_PREF))
                }
                true
            }
        initializeListSummary(findPreference<Preference>(TextSecurePreferences.NOTIFICATION_PRIVACY_PREF) as ListPreference?)
        findPreference<Preference>(TextSecurePreferences.NOTIFICATION_PRIORITY_PREF)!!.onPreferenceClickListener =
            Preference.OnPreferenceClickListener {
                val intent = Intent(Settings.ACTION_CHANNEL_NOTIFICATION_SETTINGS)
                intent.putExtra(
                    Settings.EXTRA_CHANNEL_ID, NotificationChannels.getMessagesChannel(requireContext())
                )
                intent.putExtra(Settings.EXTRA_APP_PACKAGE, requireContext().packageName)
                startActivity(intent)
                true
            }
        initializeRingtoneSummary(findPreference(TextSecurePreferences.RINGTONE_PREF))
        initializeMessageVibrateSummary(findPreference<Preference>(TextSecurePreferences.VIBRATE_PREF) as SwitchPreferenceCompat?)
    }
    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        addPreferencesFromResource(R.xml.preferences_notifications)
    }
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == 1 && resultCode == Activity.RESULT_OK && data != null) {
            var uri = data.getParcelableExtra<Uri>(RingtoneManager.EXTRA_RINGTONE_PICKED_URI)
            if (Settings.System.DEFAULT_NOTIFICATION_URI == uri) {
                NotificationChannels.updateMessageRingtone(requireContext(), uri)
                prefs.removeNotificationRingtone()
            } else {
                uri = uri ?: Uri.EMPTY
                NotificationChannels.updateMessageRingtone(requireContext(), uri)
                prefs.setNotificationRingtone(uri.toString())
            }
            initializeRingtoneSummary(findPreference(TextSecurePreferences.RINGTONE_PREF))
        }
    }
    private inner class RingtoneSummaryListener : Preference.OnPreferenceChangeListener {
        override fun onPreferenceChange(preference: Preference, newValue: Any): Boolean {
            val value = newValue as? Uri
            if (value == null || TextUtils.isEmpty(value.toString())) {
                preference.setSummary(R.string.preferences__silent)
            } else {
                RingtoneManager.getRingtone(activity, value)
                    ?.getTitle(activity)
                    ?.let { preference.summary = it }
            }
            return true
        }
    }
    private fun initializeRingtoneSummary(pref: Preference?) {
        val listener = pref!!.onPreferenceChangeListener as RingtoneSummaryListener?
        val uri = prefs.getNotificationRingtone()
        listener!!.onPreferenceChange(pref, uri)
    }
    private fun initializeMessageVibrateSummary(pref: SwitchPreferenceCompat?) {
        pref!!.isChecked = prefs.isNotificationVibrateEnabled()
    }
    private inner class NotificationPrivacyListener : ListSummaryListener() {
        @SuppressLint("StaticFieldLeak")
        override fun onPreferenceChange(preference: Preference, value: Any): Boolean {
            object : AsyncTask<Void?, Void?, Void?>() {
                override fun doInBackground(vararg params: Void?): Void? {
                    ApplicationContext.getInstance(activity).messageNotifier.updateNotification(activity!!)
                    return null
                }
            }.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR)
            return super.onPreferenceChange(preference, value)
        }
    }
    @SuppressLint("StaticFieldLeak")
    private inner class LedColorChangeListener : ListSummaryListener() {
        override fun onPreferenceChange(preference: Preference, value: Any): Boolean {
            object : AsyncTask<Void?, Void?, Void?>() {
                override fun doInBackground(vararg params: Void?): Void? {
                    NotificationChannels.updateMessagesLedColor(
                        activity!!,
                        (value as String)
                    )
                    return null
                }
            }.execute()
            return super.onPreferenceChange(preference, value)
        }
    }
    companion object {
        @Suppress("unused")
        private val TAG = NotificationsPreferenceFragment::class.java.simpleName
        fun getSummary(context: Context): CharSequence = when (isNotificationsEnabled(context)) {
            true -> R.string.ApplicationPreferencesActivity_On
            false -> R.string.ApplicationPreferencesActivity_Off
        }.let(context::getString)
    }
}